Mesh Reader Initial Implementation

zacharyknowlan opened the pull request:
Bug fixes, initial mesh reader implementation, and initial mesh reader testing.

zacharyknowlan committed:
unit_tests updated to link with Kokkos.

zacharyknowlan committed:
Mesh constructor (reader) written.

zacharyknowlan committed:
clang-format updated to remove error when committing.

zacharyknowlan committed:
Testing implemented for the mesh constructor. Test mesh files added.

Fuad-HH commented:
Looks good to me. Please test it for GPU, and then we will be good to go.

zacharyknowlan committed:
Device view added for mesh data structure.

Currently uses LayoutRight and only exists when Kokkos is compiled
with KOKKOS_ENABLE_CUDA=ON.

zacharyknowlan commented:
Device view has been created in the LayoutRight configuration which will most likely be suboptimal for performance on the GPU. This can be changed later as long as the host view is updated accordingly.

Fuad-HH commented on src/Mesh.h:
```diff
@@ -25,19 +25,43 @@ class Mesh
+         */
+        [[nodiscard]]
+        inline int GetDeviceGlobalDof(int element, int node) const {return device_data_(element,node,0);}
+    #endif
```
Instead of using two different versions of `GetCoordinates` function, use the `KOKKOS_INLINE_FUNCTION` keyword to mark it device or host function automatically. Here, you will find examples: https://kokkos.org/kokkos-core-wiki/ProgrammingGuide/ParallelDispatch.html

Fuad-HH commented on src/Mesh.h:
```diff
@@ -25,19 +25,43 @@ class Mesh
-    // data arrays depending on cuda flag
+    
+    // Mesh data in CPU memory space
+    Kokkos::View<double***, Kokkos::LayoutRight, Kokkos::HostSpace> data_; 
+    
+    // Mesh data in GPU memory space if using CUDA
+    #ifdef KOKKOS_ENABLE_CUDA
+        Kokkos::View<double***, Kokkos::LayoutRight, Kokkos::CudaSpace> device_data_; 
+    #endif
```
Instead of using `data_` and `device_data_`, Let's replace them with `host_data_` and `data_` respectively. Then, with the `KOKKOS_ENABLE_CUDA` flag you can either deep copy or just do `data_ = host_data_` which will create a view without copying. Then the user can just use `data_`.

abhiyanpaudel replied:
I think it's better to write the following way so that we won't have any layout or memory space issues:

```cpp
Kokkos::View<double***> data_(dim1, dim2, dim3);
auto host_data_ = create_mirror_view(data_);

#ifdef KOKKOS_ENABLE_CUDA
Kokkos::deep_copy(data_, host_data)
#else
data_ = host_data_
#endif
```

Fuad-HH reviewed (commented):
We will need this when we will be testing on GPU

zacharyknowlan committed:
Updates to Mesh class for device execution.

By default data_ now exists on the device.
Testing updated to copy data_ before testing on the CPU.

Fuad-HH commented:
All looks good to me. @abhiyanpaudel, @zacharyknowlan, Should we merge this now?

zacharyknowlan commented:
I think it can be merged as long as @abhiyanpaudel is good with the current layout and format of the mesh data.

Fuad-HH commented:
I am merging it now. We can add something later if we need to.
